### PR TITLE
[catmem] Release Queue Descriptor Even If `close()` Fails

### PR DESCRIPTION
**Pull request by ppenna**
## Description

This PR is a follow up of https://github.com/demikernel/demikernel/pull/686

## Summary of Changes

- Changed `close()` to release the queue descriptor even if we fail to `push_eof()`.

https://github.com/demikernel/demikernel/pull/686

### CODE REVIEW

File: src/rust/catmem/mod.rs [hunk 1]
```diff
@@ -167,17 +167,27 @@ impl CatmemLibOS {
...
+        // Check if queue descriptor is valid.
+        if qtable.get(&qd).is_none() {
+            let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
+            error!("close(): {}", cause);
+            return Err(Fail::new(libc::EBADF, &cause));
+        }
+
+        // Attempt to push EoF.
+        let result: Result<(), Fail> = {
+            // It is safe to call expect() here because the queue descriptor is valid guaranteed to be valid.
```
Comment (iyzhang) [inline src/rust/catmem/mod.rs:181]: take out "is valid"

Comment (iyzhang) [inline src/rust/catmem/mod.rs:181]: take out "is valid"

**Review by iyzhang — APPROVED** [review]
LGTM

**Comment by iyzhang** [comment]
instead of freeing the queue descriptor, should we eventually have states and the queue should be placed into a CLOSING state, which continues to try to push the EoF?

**Comment by ppenna** [comment]
> instead of freeing the queue descriptor, should we eventually have states and the queue should be placed into a CLOSING state, which continues to try to push the EoF?

Definitely we should have some abstraction for the states of the queue itself, such as we have in Catnap Loop. However,  the problem here about continuing try to push EoF is that we are not sure that this is due because the ring buffer is full or the other end has closed the pipe. In the latter case, we don't have a signal mechanism, because a pipe is unidirectional. Any ideias?

File: src/rust/catmem/mod.rs (hunk 1 above)
Comment (ppenna) [reply src/rust/catmem/mod.rs:181]: Sure 😄

**Comment by iyzhang** [comment]
> > instead of freeing the queue descriptor, should we eventually have states and the queue should be placed into a CLOSING state, which continues to try to push the EoF?
> 
> Definitely we should have some abstraction for the states of the queue itself, such as we have in Catnap Loop. However, the problem here about continuing try to push EoF is that we are not sure that this is due because the ring buffer is full or the other end has closed the pipe. In the latter case, we don't have a signal mechanism, because a pipe is unidirectional. Any ideias?

My thought was that we should at least keep trying to push EoF regardless of whether the other end will read it because it closed the pipe or stopped pulling things out of it.

**Comment by ppenna** [comment]
> > > instead of freeing the queue descriptor, should we eventually have states and the queue should be placed into a CLOSING state, which continues to try to push the EoF?
> > 
> > 
> > Definitely we should have some abstraction for the states of the queue itself, such as we have in Catnap Loop. However, the problem here about continuing try to push EoF is that we are not sure that this is due because the ring buffer is full or the other end has closed the pipe. In the latter case, we don't have a signal mechanism, because a pipe is unidirectional. Any ideias?
> 
> My thought was that we should at least keep trying to push EoF regardless of whether the other end will read it because it closed the pipe or stopped pulling things out of it.

Sure, I've opened an issue for this.

**Pull request merged**